### ISSUE TITLE
Addition of generic ICloneable interface

### ISSUE THREAD

**Issue by Hawkeye4040** [dotnet/runtime#49433]
## Background and Motivation

The initial motivation for this API proposal was the common use of Clone Methods that return a specific type.

## Proposed API

```C#
namespace System
{
    public interface ICloneable<T>
    {
        T Clone();
    }
}
```
## Usage Examples

Single implementation of generic cloning

```C#

    
    public class MyClass : ICloneable<MyClass>
    {
         public string Name { get; set; }

         public MyClass Clone()
         {
             return new MyClass{ Name = Name };
         }
    }
```
Implementation of generic and object cloning using overloads

```C#
   public class MyClass : ICloneable, ICloneable<MyClass>
    {
         public string Name { get; set; }
         
         public object Clone()
         {
             return (object)Clone();
         }

         public MyClass Clone()
         {
             return new MyClass{ Name = Name };
         }
    }
```
Generic Cloning and explicit implementation of object cloning

```C#
    
    public class MyClass : ICloneable, ICloneable<MyClass>
    {
         public string Name { get; set; }
         
         object ICloneable.Clone() => Clone();

         public MyClass Clone()
         {
             return new MyClass{ Name = Name };
         }
    }
```

## Alternative Designs

I originally considered putting this in one of my "Core" libraries, however without it being implemented in .NET directly I'd end up re-using it or making a ".NET extensions" library anyway. The latter seems especially redundant now that .NET is open source.

## Risks

No known risks.

**Comment by jkotas** [comment]
Duplicate of #37561

**Comment by Hawkeye4040** [comment]
Oh nice I did look already I must've missed it.

**Comment by Hawkeye4040** [comment]
### **_I completely disagree with the conclusions drawn in Issue [#37561](https://github.com/dotnet/runtime/issues/37561)._** 

Because three people think that the original ICloneable interface was a mistake, something that wouldn't negatively impact anything gets turned down?

 Just because it's convenient but serves no real functional change that they see doesn't necessarily mean there isn't one first off, and second if we aren't adding anything just for the sake of convenience we should then ditch LINQ and most of the more recent C# updates then. Let's write everything out the hard way it how that so called logic comes off to me. 
 
I still propose this should be added. I see many people using strongly typed cloning without the implementation of an interface. Rather than us having to all put that in some core library that we re-use in our apps constantly or have to constantly re-write that why not just put it into the .NET Runtime itself?

Rather than turning down useful additions why not focus on fixing actual issues?

**Comment by danmoseley** [comment]
@Hawkeye4040, thanks for your proposal. At the same time, please [assume good intent](https://github.com/dotnet/runtime/blob/main/CODE-OF-CONDUCT.md).

As noted there, if we add an interface, the contract must be clear (not just the API shape). Based on experience with the existing interface, is not clear that there is good contract that could be widely agreed on. At the same time, it is not clear that the problems can't be solved by creating your own interface(s).

Finally, any API we add has a cost in terms of cognitive load amongst other things. If we have this interface, should we use it in existing types? New types? What guidance would we document around it? Would it lead folks to adopt bad patterns, etc. If you create your own interface, it sidesteps some of those issues.

**Comment by GrabYourPitchforks** [comment]
With new versions of C#, there's a really neat feature that obviates the need for `ICloneable<T>`. You can now specify _more derived types_ when overriding virtual members. Here's an example:

```cs
using System;

public class C {
    public void M() {
        BaseClass class1 = new BaseClass();
        BaseClass cloned1 = class1.Clone();
        
        DerivedClass class2 = new DerivedClass();
        DerivedClass cloned2 = class2.Clone();
        
        BaseClass castClass2 = class2;
        BaseClass clonedClass2 = castClass2.Clone();
        Console.WriteLine(clonedClass2.GetType()); // 'typeof(DerivedClass)'
    }
}

public class BaseClass : ICloneable
{
    public virtual BaseClass Clone()
    {
        return new BaseClass();
    }
    
    object ICloneable.Clone() => Clone();
}

public class DerivedClass : BaseClass
{
    public override DerivedClass Clone()
    {
        return new DerivedClass();
    }
}
```

With this pattern, regardless of what your local variable is typed as, you're free to call the `Clone` method and get back a strongly-typed instance without requiring the caller to perform an explicit cast. The exception is that if all you have is an instance of _ICloneable_, you have to explicit cast to the type you expect, but that's no different than any other usage of _ICloneable_ today.

**Comment by Hawkeye4040** [comment]
Except for when you're trying to override it in the implementing class....

Again, just because you personally don't see a need, doesn't mean there isn't one. Perhaps someone who works in a totally different field sees something you would never have thought of. That is a very sloppy assumption, it would be as if saying I know the entire scale and scope of everything that will ever be programmed and I know for a fact you'll never need this.

> With this pattern, regardless of what your local variable is typed as, you're free to call the Clone method and get back a strongly-typed instance without requiring the caller to perform an explicit cast. The exception is that if all you have is an instance of ICloneable, you have to explicit cast to the type you expect, but that's no different than any other usage of `ICloneable` today.

Will all due respect there is nothing remotely accurate about that statement. Stating things in a matter of fact tone doesn't automatically make you correct.

There are many reasonable usages of a generic cloning interface. It is just very evident you don't see those usages yourself personally. That is fine, then simply just don't use it. Adding it does absolutely nothing to you. It changes nothing (If done correctly of course that should be a given but even with this here someone will say it). That being said, not having this affects many of us that do want it. So I have to say it blows my mind a little bit to see so many people vehemently protest such a thing, going out of their way to hold others back. If there is **actually** a constructive reason here for this I'm all ears but I have yet to hear one. Just nonsense that isn't even remotely accurate.

In summary, I feel as though the opposition to this addition just need to stop going out of their way to limit others for no good reason. If it doesn't affect you at all why come ruin it for those of us that do see the need for it? It won't change anything for you.

**Comment by Hawkeye4040** [comment]
Instead of down thumbing me which is why I'm so irritated by this to begin with why not constructively discuss this? Just like one person did already at least.

Okay @danmoseley I see your point here with your concerns. I don't think any of the issues you brought up such as it leading to bad patterns will come up. I can see why that's a concern. I've seen Generic Cloning proposed in some very poor ways before. Perhaps this explains some of the ambivalence here. 

The way in which I'm proposing it shouldn't even require updates to source or documentation nor would it break anything. It doesn't also require the implementation of the original `ICloneable` I've seen that proposed which is a bad idea. I've also seen it proposed with needless constraints. 

On those I would **completely** agree, however the manner in which I am proposing it be implemented either addresses or nullifies all those issues, at least the ones I know about. With that being said every last argument against this proposal would be true of those other proposals but not this one.

**Comment by GrabYourPitchforks** [comment]
> Except for when you're trying to override it in the implementing class....

Not sure what you mean. The code sample I provided above showed that you can override a method and change the return type of the method at the same time that you override it. Ideally this would serve as a good replacement for any proposed `ICloneable<T>` interface.

Let's say we added an `ICloneable<T>`. How do you see people using that interface - both as a consumer of the interface and as an implementer of the interface? What would the sample code look like? Those kinds of examples would help us make a more informed decision.

**Comment by Hawkeye4040** [comment]
Very well. I'm sure I can dig up some better one's too that I've seen and I know I have some myself.

To clarify on the overriding it in an implementing class I think my point is best illustrated by the scenario of a sealed class implementing the genric interface and not necessarily the original `ICloneable` interface.

Comparison of interfaces (to be explicitly clear not that the generic version doesn't require implementation of the original I've seen that proposed before which I disagree with). 

Due to that fact you wouldn't have to update anything (including doc comments that was brought up before) with existing implementations of `ICloneable` _**unless**_ you also implement `ICloneable<T>`.

```C#
public interface ICloneable
{
    object Clone();
}

public interface ICloneable<T>
{
    T Clone();
}
```

Sealed Class Scenario:

```C#
using System;

public sealed class SealedExampleClass : ICloneable<SealedExampleClass>
{
    public SealedExampleClass Clone()
    {
        // Clone here.
    }
}
```

This allows for various additional scenarios I've seen without any (known) conflicts and I've personally analyzed a great many of them. It seems as though many opposing this are assuming that they'd always need to be implemented together although I'm not entirely sure on that one. As this explicit implementation example shows they both can be but don't have to be.

```C#

using System;

public sealed class SealedExampleClass : ICloneablle, ICloneable<SealedExampleClass>
{

  object ICloneable.Clone() => Clone();

  public SealedExampleClass Clone()
  {
      // Clone here.
  }
}

```

I've seen many more examples I'm sure I could dig more up. These were just off the top of my head but the point is this allows you to use various patterns and gives you the freedom to pretty much do it any way you would need to. I haven't seen a scenario this doesn't work for yet. If you can find one I'd love to see it and pick it apart though.

**Comment by tfenise** [comment]
Why does the class `SealedExampleClass` need to implement `ICloneable<SealedExampleClass>` or `ICloneable` or any similar interface? Without implementing any interface, the class still has a method `Clone()` and users are still able to call it. Is there any scenario where implementing such interface makes a difference? Such as some method having a parameter with type `ICloneable` or `ICloneable<T>`, or returing `ICloneable` or `ICloneable<T>`, or a generic constraint `where T : ICloneable<T>`, or like `IDisposable` and `using(...)`, some special language construct around `ICloneable`?

**Comment by Hawkeye4040** [comment]
Well then by that logic why do we ever implement any interface at all?

 This was an example presented for another reason it was not considerate of what you're talking about here and therefore a null point.

`SealedExampleClass`was just to illustrate a concept to answer an unrelated question not serve as a complete implementation of anything.

**Comment by Hawkeye4040** [comment]
In addition I'd argue against using the constraint. A lot of those against this interface use that as a major con against this interface. 

Keeping it more open ended (see my initial examples) would mean there wouldn't be issues with broken functionality, doc updates and various other issues mentioned before.

**Comment by tfenise** [comment]
> Well then by that logic why do we ever implement any interface at all?

You can do `foreach` with an `IEnumerable<T>`. You can do `using` with a `IDisposable`. You can call `Enumerable.ToArray()` on an `IEnumerable<T>`. That's why those interfaces are useful.

Then what is the use for `ICloneable<T>`? If one need to call `ICloneable<T>.Clone()`, they probably already know the object is of the concrete type `T` (as long as they don't use a generic constraint `where T : ICloneable<T>`, they know the exact type which `T` refers to), then even if there is no `ICloneable<T>`, users can still call `T.Clone()` defined on the class `T`. Adding an `ICloneable<T>` does not make their life easier.

This is not the case for `IEnumerable<T>`. Someone calling `Enumerable.ToArray()` on an `IEnumerable<T>` may not know whether the `IEnumerable<T>` is a `T[]` or `List<T>` or whatever.

In your first example, `class MyClass : ICloneable<MyClass>`. If one need to call `ICloneable<MyClass>.Clone()`, they probably already know it is a `MyClass`, and can call `MyClass.Clone()`, even without the interface.

What your examples show are possible implementations of `ICloneable<T>`. Is there any use case of `ICloneable<T>`?

> As I've said many times on this thread just because you don't yourself see a use doesn't mean there isn't one. It also doesn't mean there aren't many of us who could make use of it so I really don't see the vehement opposition to this.
> 
> If you don't want to use it simply don't and it won't affect you. Why go out of your way to hold others back for no reason?

If you think there is a use, please provide an example _using_ `ICloneable<T>`, not just _implementing_ `ICloneable<T>`.

**Comment by Hawkeye4040** [comment]
Please read the whole thread. I'm tired of repeating myself.

**Comment by GrabYourPitchforks** [comment]
> Please read the whole thread.

No example of *using* this interface was ever given in any sample code. The question remains: if `ICloneable<Foo>` is only ever implemented by class `Foo`, then as a consumer when would I ever have a local / field / parameter of type `ICloneable<Foo>`? What benefit would it provide over having a local / field / parameter of type `Foo` directly?

That's the part we're all confused about. Without concrete code samples showing why this is useful we can't make forward progress on it.

**Comment by Hawkeye4040** [comment]
Let me dig up something that I can post or re-write it. 

I suppose that's the only way to illustrate it better for you. As the Foo Bar type examples are proving inadequate here. I personally thought it illustrated the point fine but perhaps it's because I've seen it used this way a lot.

This one may take me a bit. Thank you for at least entertaining the notion though.

**Comment by Hawkeye4040** [comment]
@GrabYourPitchforks First off I honestly thought it was more obvious just with the implementation shown. It honestly seemed like everyone was being a _jerk_ about a good suggestion. Well truthfully there are some doing that and it isn't appreciated. I cannot fathom why you're going out of your way to give me your thumbs down on a topic that doesn't affect you. Don't you have some work to do? I'm going to tell your boss your just sitting on GitHub trolling people all day. Not you pitchforks they know who they are though. I don't think the community guidelines covered flaming but I believe it was **_heavily_** implied. Not cool.

Anyway, had I originally thought more elaboration was needed I certainly would have. I'm sure that contributed to the confusion. I hope I follow you right with your question about what benefit it provides.

I wish I could just show you a certain project but I might just have to Foo Bar it. I've personally encountered scenarios where having a return type of object instead of something more explicit causes many issues when this API is consumed I find a significant number of try-catch loops and checks for null are needed. I know there's a lot of cases where you can just cast to Foo from object but if Bar is passed instead in some cases I have a massive amount of checks to write that could be resolved with 15 simple lines (including comments). 

Currently this _**exact**_ implementation I'm proposing sits in my organization's core library and it works wonderfully. There's a couple scenarios that I know I've encountered myself as well as seen in open source somewhere where overloads and explicit implementation as shown in my original comment came up and this was actually the path of least resistance. There is at least one known case using reflection where having the explicit type returned from a clone method will matter. I'm sure there's more but I can't confirm those for you yet.

Even if you don't add it I will still of course continue to use it with great success. I feel it would be a shame, however, not to see it in .NET itself for others because I know I'm not the only person who has encountered this need.

**Comment by jkotas** [comment]
[.NET Foundation Code of Conduct](https://dotnetfoundation.org/about/code-of-conduct) prohits personal attacks that your last comment comes very close to. Please keep the discussion focused on the technical matter.

Github reactions are a simple and effective way to express ones technical opinion on issues, PR and comments. There is nothing wrong in using them extensively. Feel free to use them extensively as well.

The problems with `ICloneable` are well understood by the .NET design team that @GrabYourPitchforks  is member of. I am sure you would get similar questions from other .NET design team members. There is even a chapter on `ICloneable` in the [.NET Framework Design Guidelines](https://learning.oreilly.com/library/view/framework-design-guidelines/9780135896457/) book. Quoting from this chapter:

> The moral of the story is that you should never ship an interface if you don’t have both implementations and consumers of the interface

It is why we are asking you to show good examples of both the consumer and the producer code. The examples mentioned here so far do not demonstrate the use of this interface well. I am sorry that we do not understand how this can work well as general purpose interface. 

> if you don't add it I will still of course continue to use it with great success. 

Yes, that's perfectly fine. There are many abstractions that people find useful, define locally and use successfully, but that are not good candidates for inclusion in the core framework.

**Comment by Hawkeye4040** [comment]
Well I'm done here. I'd expect better from Microsoft employees. This is obnoxious. I'll be speaking to your superiors about your conduct here.
